modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/streaming/PagingProviderProducer.java

```diff
@@ -77,16 +89,37 @@ public int getSize() {
    * @return
    */
   private <R> R performWithConnection(Function<Object, R> function) {
-    ConnectionSupplier connectionSupplier = null;
+    Optional<MutableConfigurationStats> stats = getMutableConfigurationStats(executionContext);
+    RetryPolicyTemplate retryPolicy =
+        (RetryPolicyTemplate) executionContext.getRetryPolicyTemplate().orElseGet(NoRetryPolicyTemplate::new);
+    CompletableFuture<R> future = retryPolicy.applyPolicy(() -> completedFuture(withConnection(function)),
+                                                          e -> shouldRetry(e, executionContext),
+                                                          e -> {
+                                                          },
+                                                          e -> stats.ifPresent(s -> s.discountInflightOperation()),
+                                                          identity(),
+                                                          executionContext.getCurrentScheduler());
+    try {
+      return future.get();
+    } catch (Exception e) {
+      throw new MuleRuntimeException(createStaticMessage(COULD_NOT_RECONNECT), e);
+    }
+  }
+
+  private <R> R withConnection(Function<Object, R> function) {
+    ConnectionSupplier connectionSupplier;
     try {
       connectionSupplier = connectionSupplierFactory.getConnectionSupplier();
-      return function.apply(connectionSupplier.getConnection());
-    } catch (MuleException e) {
-      throw new MuleRuntimeException(createStaticMessage(COULD_NOT_OBTAIN_A_CONNECTION), e);
-    } finally {
-      if (connectionSupplier != null) {
+      try {
```

Comment:
Is it necessary to have nested try-catch scopes? Can we make it simpler?

Reply:
Simplified it.